ibm/data_source_ibm_appid_roles.go

```diff
@@ -0,0 +1,117 @@
+package ibm
+
+import (
+	"context"
+	"fmt"
+	appid "github.com/IBM/appid-management-go-sdk/appidmanagementv4"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"sort"
+)
+
+func dataSourceIBMAppIDRoles() *schema.Resource {
+	return &schema.Resource{
+		Description: "A list of AppID roles",
+		ReadContext: dataSourceIBMAppIDRolesRead,
+		Schema: map[string]*schema.Schema{
+			"tenant_id": {
+				Description: "The AppID instance GUID",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"roles": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"role_id": {
+							Description: "Role ID",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"name": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Unique role name",
+						},
+						"description": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Optional role description",
+						},
+						"access": {
+							Type:     schema.TypeSet,
+							Computed: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"application_id": {
+										Type:     schema.TypeString,
+										Computed: true,
+									},
+									"scopes": {
+										Type: schema.TypeList,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
+										Computed: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceIBMAppIDRolesRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	appIDClient, err := meta.(ClientSession).AppIDAPI()
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tenantID := d.Get("tenant_id").(string)
+
+	roles, _, err := appIDClient.ListRolesWithContext(ctx, &appid.ListRolesOptions{
+		TenantID: &tenantID,
```

Comment:
Do we. need any pagination here because many ibmcloud API with List just list only first page

Reply:
Hm good question, but I don't see any pagination parameters for roles in their swagger spec, I will check with team